### ISSUE TITLE
Compare runtimes of spaCy NER pipelines using CPU and GPU

### ISSUE THREAD

**Issue by FrancescoCasalegno** [BlueBrain/Search#337]
## Description

While adopting a `transformer` backbone for our `spaCy` NER models may be beneficial in terms of accuracy (see #335), this may also imply slower runtime with respect to using a simpler `tok2vec`.

To run on GPUs using `spaCy` it seems that only 2 things are needed (see [here](https://spacy.io/usage#gpu) for complete guide).
 1. `pip install --upgrade spacy[<cuda_version>]`;
 2. specify [`spacy.require_gpu()`](https://spacy.io/api/top-level#spacy.require_gpu) before any `spacy.load(some_model)`.

The GPU version should be checked before installing the right `spacy[<cuda_version>]`. For instance, given
```bash
$ nvcc --version
nvcc: NVIDIA (R) Cuda compiler driver
Copyright (c) 2005-2019 NVIDIA Corporation
Built on Wed_Oct_23_19:24:38_PDT_2019
Cuda compilation tools, release 10.2, V10.2.89
```
we should `pip install --upgrade spacy[cuda102]`, since we have `cuda v10.2`.

## Actions
Collect results of runtimes of inference using `spaCy` pipelines in a variety of settings:
   - [x] running on  CPU vs GPU
   - [x] using backbone `tok2vec` vs `transformer`
   - [x] with `pipeline = ["transformer", "ner"]` vs `pipliene = ["transformer", "tagger", "attribute_ruler", "lemmatizer", "parser", "ner", "entity_ruler"]`

**Comment by EmilieDel** [comment]
First results:
- Experiment 1:
  - Pipeline: ['tok2vec', 'tagger', 'attribute_ruler', 'lemmatizer','parser', 'ner']
  - Frozen components:  ['tok2vec', 'tagger', 'attribute_ruler','lemmatizer', 'parser']
  - 50 epochs (no early stopping)
  - Runtime:
    - **CPU: 93.5 +/- 1.5 sec** (on 2 experiments)
    - **GPU: 124 +/- 1 sec** (on 2 experiments)
    - Training on GPU is ~1,3x slower than on CPU for this configuration
- Experiment 2:
  - Pipeline:  ['transformer', 'ner']
  - Frozen components: []
  - 50 epochs (no early stopping)
  - Runtime:
    - **CPU: 1214.5 +/- 22.5 sec** (2 experiments)
    - **GPU:  173.75 +/- 13.5** (4 experiments - side note: training seems not reproducible, losses are different each time).
    - Training on GPU is ~7x faster than on CPU for this configuration

**Comment by EmilieDel** [comment]
## Results of Inference runtime
### Experiments:
- Model: `en_core_web_lg` (=`tok2vec` backbone) / `en_core_web_trf` (=`transformers` backbone)
- Device: CPU / GPU
- Pipeline: `['backbone', 'ner']` (referred as `ner`) / `['backbone', 'tagger', 'parser', 'attribute_ruler', 'lemmatizer', 'ner', 'entity_ruler']` (referred as `all`)
- Sentence: 0 / 10 / 100 spans detected

### Runtime:
![Screenshot 2021-04-26 at 15 15 53](https://user-images.githubusercontent.com/47669575/116088773-78691f00-a6a2-11eb-8239-02478e2b1c89.png)
![Screenshot 2021-04-26 at 15 15 59](https://user-images.githubusercontent.com/47669575/116088771-77d08880-a6a2-11eb-939e-4b6d3e16af80.png)
![Screenshot 2021-04-26 at 15 16 12](https://user-images.githubusercontent.com/47669575/116088767-7737f200-a6a2-11eb-9f95-73b1815056fc.png)

**Comment by jankrepl** [comment]
Thank you for the great work!
As discussed during the sprint review, just 2 things (not necessarily actionable) to keep in mind related to your benchmarks vs our production code.
* We are actually using `nlp.pipe`
* We are heavily relying on multiprocessing when to create the mining cache and generalizing to work with multiple GPUs could be tricky

**Comment by FrancescoCasalegno** [comment]
Thank you @EmilieDel for your great work!

Related to @jankrepl's comment above, I would have 2 requests before I think this ticket can be considered Done.
- [ ] Run same experiments with [`nlp.pipe`](https://spacy.io/api/language#pipe) and `texts` (i.e. the `Iterable` that is consumed) of varying length—e.g. use generators that yield 10, 100, 10,000, 100,000 items.
- [ ] Report here all results (both the ones you already did + the ones with `nlp.pipe`) in a table in text format.

Example of such a table:

| Col1  | Col2 |
| ------------- | ------------- |
| 1.234  | 2.123  |
| 2.331  | 2.312  |
| 1.111  | 2.121  |

**Comment by EmilieDel** [comment]
Here are the results for `nlp.pipe` and `texts` being 10, 100, 10,000 and 100,000 sentences:

### Results:
Summary Table:
![Screenshot 2021-04-27 at 18 04 00](https://user-images.githubusercontent.com/47669575/116275174-a96d5080-a783-11eb-93db-2aabf12e2afd.png)
Histograms:
![Screenshot 2021-04-27 at 18 08 21](https://user-images.githubusercontent.com/47669575/116275220-b5f1a900-a783-11eb-93dc-5befadec32f2.png)
![Screenshot 2021-04-27 at 18 07 54](https://user-images.githubusercontent.com/47669575/116275224-b722d600-a783-11eb-8919-bac1050bbdfd.png)
![Screenshot 2021-04-27 at 18 06 42](https://user-images.githubusercontent.com/47669575/116275225-b722d600-a783-11eb-97c7-6cce3d3f47de.png)
![Screenshot 2021-04-27 at 18 06 35](https://user-images.githubusercontent.com/47669575/116275226-b722d600-a783-11eb-8990-a72828e732c7.png)


### Observation:
- There is still some differences between CPU and GPU but the runtime is so small that it is not really a problem in my opinion! 
- Differences between CPU and GPU seems to fluctuate more than in the previous experiment.

**Comment by EmilieDel** [comment]
# Update: Results of the inference runtime
## Experiments
### Features
  - Model: `en_core_web_lg` (=`tok2vec` backbone) / `en_core_web_trf` (=`transformers` backbone)
  - Device: `CPU` / `GPU`
  - Pipeline: `['backbone', 'ner']` (referred as `ner`) / `['backbone', 'tagger', 'parser', 'attribute_ruler', 'lemmatizer', 'ner', 'entity_ruler']` (referred as `all`)
### First experiment: `nlp(sentence)`
  - Sentence: 0 / 10 / 100 spans detected
### Second experiment: `nlp.pipe(texts, as_tuples=True)`
  - Number of Sentences: 1 / 10 / 100 

## Results
### First experiment

mode | tok2vec_ner | tok2vec_all | transformer_ner | transformer_all
-- | -- | -- | -- | -- 
100_spans_GPU | 16.423931 | 36.083158 | 32.687129 | 46.542662
10_spans_GPU | 11.148465 | 14.184615 | 17.283405 | 19.916191
0_spans_GPU | 11.149544 | 13.708422 | 16.170592 | 19.447631
100_spans_CPU | 21.721511 | 41.319433 | 159.299527 | 181.267698
10_spans_CPU | 5.862713 | 8.343925 | 43.514955 | 44.946311
0_spans_CPU | 5.529749 | 7.552872 | 49.984276 | 50.154468

![Screenshot 2021-04-28 at 12 28 25](https://user-images.githubusercontent.com/47669575/116389485-462ffc80-a81d-11eb-8b36-4c76a1afa95d.png)

### Second experiment


mode | tok2vec_all | tok2vec_ner | transformer_all | transformer_ner
-- | -- | -- | -- | --
1_sentence_GPU | 15.038188 | 12.677844 | 22.005937 | 16.799952
10_sentences_GPU | 31.251985 | 20.575479 | 48.117278 | 32.152022
100_sentences_GPU | 211.672902 | 100.827304 | 551.115255 | 382.475122
1_sentence_CPU | 7.017947 | 5.325186 | 51.916081 | 50.175784
10_sentences_CPU | 24.786860 | 16.008097 | 152.222355 | 140.299771
100_sentences_CPU | 264.803753 | 159.031736 | 1981.153947 | 1917.262415


![Screenshot 2021-04-28 at 14 40 14](https://user-images.githubusercontent.com/47669575/116405286-c3fd0380-a82f-11eb-98b1-846136baa674.png)

## Code
```python
"""Code to reproduce experiment."""
import timeit

import numpy as np
import spacy
import sqlalchemy
import torch.multiprocessing as mp

from bluesearch.sql import retrieve_sentences_from_sentence_ids

mp.set_start_method('spawn')  # for n_process is bigger than 1 

# 1. Initiate spacy model
model = "en_core_web_trf"          # or "en_core_web_lg"
pipes_to_disable = []              # or ['tagger', 'parser', 'attribute_ruler', 'lemmatizer']
entity_ruler = True                # or False
DB_URL = ""                        # URL to the database

spacy.require_gpu()                # or spacy.require_cpu()
nlp = spacy.load(model, disable=pipes_to_disable)
if entity_ruler:
    ruler = nlp.add_pipe("entity_ruler")
    ruler.add_patterns([{"label": "LOCATION", "pattern": "Berlin"}])

# 2. First experiment: nlp(sentence)
# 2.a. Create sentence to process
n_times = 1
sentence = "Emmanuel Macron is living in Paris." * n_times
# 2.b. Results
doc = nlp(sentence)
print(doc.ents)
# 2.b. Compute runtime
time_results = timeit.repeat(lambda: nlp(sentence), repeat=10, number=10)
mean_time = np.mean(time_results)
print(mean_time)

# 3. Second experiment: nlp.pipe(texts)
# 3.a. Load sentences from database
engine = sqlalchemy.create_engine(DB_URL)
n_sentences = 10
n_process = 1 
sentences = retrieve_sentences_from_sentence_ids(np.arange(1, n_sentences+1), engine)
texts = [(sent["text"], {"article_id": sent["article_id"], "sentence_id": sent["sentence_id"]}) for i, sent in sentences.iterrows()]
all_texts.append(texts)
# 3.b. Compute runtime
time_results = timeit.repeat(lambda: list(nlp.pipe(texts, as_tuples=True)), repeat=10, number=10)
mean_time = np.mean(time_results)
print(mean_time)
```

**Comment by FrancescoCasalegno** [comment]
Thanks @EmilieDel for these experiments!

Having a look at the ratio between CPU runtime vs. GPU runtime for the same test case can also be interesting.
<img src="https://user-images.githubusercontent.com/17013890/116417670-cb2a0e80-a83b-11eb-8c0e-2ec89badfe0a.png" width=500px>
<img src="https://user-images.githubusercontent.com/17013890/116417671-cc5b3b80-a83b-11eb-898f-929807ce93c8.png" width=500px>

@pafonta you mentioned this morning that training with GPU was ~4x faster, which seems to be consistent with these results right?

**Comment by pafonta** [comment]
Hello @FrancescoCasalegno !

> @pafonta you mentioned this morning that training with GPU was ~4x faster, which seems to be consistent with these results right?

Inference and training times of a Transformer-based NER-only pipeline seem indeed to be ~4x faster on GPU than on CPU.

(source: https://github.com/BlueBrain/Search/issues/335#issuecomment-828249087)

**Comment by EmilieDel** [comment]
Hello @pafonta, @FrancescoCasalegno, here are some analysis for `nlp.pipe()`. 

### Setup and Features
- The number of sentences was fixed to 100.000 sentences for every experiment. 
- Each experiment was ran once. 
- The model used was `en_core_web_lg` (which means `tok2vec` for the backbone) (WIP for `en_core_web_trf`)
- Device: CPU / GPU
- Multiprocessing: `spawn` / `fork`
- Batch size: 256 / 500 / 1000
- N processes: 1 / 2 / 3 

### Results
![image](https://user-images.githubusercontent.com/47669575/116431111-bfdce000-a847-11eb-8e88-b30149fe7102.png)

![image](https://user-images.githubusercontent.com/47669575/116431180-d2571980-a847-11eb-95a3-9cd169b9b52b.png)

### Observation
- It seems 2 processes for CPU are enough to be quicker for than 1 process for GPU. However, having a bigger number of  sentences is already reducing runtimes between CPU and GPU.
- Increasing the `batch_size` is helping when multiprocessing mode is `spawn` but not when it is `fork`. (observed even when `n_process=1`)

**Comment by FrancescoCasalegno** [comment]
My two cents.

1. The speedup factor is only a half of the elements we need to look at to make the decision on CPU vs GPU. The other half is: what is the runtime on CPU? Because even if the speedup factor is not huge we may still consider the runtime on CPU as acceptable.
2. To decide on CPU vs GPU we also need to look at the reproducibility Issue #343 
3. Here we are showing results on **inference**, so results on **training** may vary. Also, is it even possible to set `n_processes` also for `spacy train` in the `config.cfg`?

**Comment by EmilieDel** [comment]
As pointed out during planning of 4th May, there was a mismatch of colors in the second plot of [this comment](https://github.com/BlueBrain/Search/issues/337#issuecomment-828565415). Here are the new results:
![Screenshot 2021-05-04 at 18 38 23](https://user-images.githubusercontent.com/47669575/117038586-19da1b80-ad08-11eb-8673-fe5e9057b543.png)

**Comment by saitej123** [comment]
Please share code for faster inference in cpu
